File: Assets/Animations/SlideAnimation.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ClassExtends = ____lualib.__TS__ClassExtends
local ____exports = {}
local ____Animation = require("Assets.Animations.Animation")
local Animation = ____Animation.default
____exports.default = __TS__Class()
local SlideAnimation = ____exports.default
SlideAnimation.name = "SlideAnimation"
__TS__ClassExtends(SlideAnimation, Animation)
function SlideAnimation.prototype.____constructor(self, gameManager, id, animDuration, offsetX, offsetY, assets, constructionOptions)
    Animation.prototype.____constructor(
        self,
        gameManager,
        id,
        assets,
        constructionOptions
    )
    self.animOffsetX = 0
    self.animOffsetY = 0
    self.animTargetOffsetX = 0
    self.animTargetOffsetY = 0
    self.animTargetOffsetX = offsetX
    self.animTargetOffsetY = offsetY
    self.animDuration = animDuration
end
function SlideAnimation.prototype.updateAnimation(self, deltaTime)
    Animation.prototype.updateAnimation(self, deltaTime)
    if not self.isAnimating then
        self:updateX(self.animTargetOffsetX)
        self:updateY(self.animTargetOffsetY)
        return
    end
    self:calculateAnimationOffset()
    self:updateX(self.animOffsetX)
    self:updateY(self.animOffsetY)
end
function SlideAnimation.prototype.calculateAnimationOffset(self)
    local progress = self.animElapsed / self.animDuration
    self.animOffsetX = self.animTargetOffsetX * progress
    self.animOffsetY = self.animTargetOffsetY * progress
end
return ____exports

File: Assets/Animations/SlideAnimation.ts
import GameManager from "GameManager";
import Animation, { AnimationAssets, AnimationOptions } from "./Animation";

export interface SlideOptions extends AnimationOptions {
  readonly animDuration?: number;
  readonly bounceEffect?: boolean;
}

export default class SlideAnimation extends Animation {
  animOffsetX: number = 0; // Current animation offset
  animOffsetY: number = 0; // Current animation offset
  animTargetOffsetX: number = 0; // Target animation offset
  animTargetOffsetY: number = 0; // Target animation offset (e.g., -20 for up)
  animDuration: number;
  bounceEffect: boolean;

  constructor(
    gameManager: GameManager,
    id: string,
    animDuration: number,
    offsetX: number,
    offsetY: number,
    assets: AnimationAssets[],
    constructionOptions?: SlideOptions
  ) {
    super(gameManager, id, assets, constructionOptions);
    this.bounceEffect = constructionOptions?.bounceEffect ?? false;
    this.animTargetOffsetX = offsetX + this.getOvershootAmount(offsetX);
    this.animTargetOffsetY = offsetY + this.getOvershootAmount(offsetY);
    this.animDuration = animDuration;
  }

  updateAnimation(deltaTime: number): void {
    super.updateAnimation(deltaTime);
    if (!this.isAnimating) {
      // finish sending the assets to their target positions
      this.updateX(this.animTargetOffsetX);
      this.updateY(this.animTargetOffsetY);
      this.applyBounceEffect();
      return;
    }

    // Interpolate animation offset
    this.calculateAnimationOffset();

    this.updateX(this.animOffsetX);
    this.updateY(this.animOffsetY);
  }

  calculateAnimationOffset(): void {
    const progress = this.animElapsed / this.animDuration;
    this.animOffsetX = this.animTargetOffsetX * progress;
    this.animOffsetY = this.animTargetOffsetY * progress;
  }

  applyBounceEffect(): void {
    if (!this.bounceEffect) return;

    // Bounce back from overshoot to actual target
    const bounceAmplitudeX = -this.getOvershootAmount(this.animTargetOffsetX);
    const bounceAmplitudeY = -this.getOvershootAmount(this.animTargetOffsetY);

    this.gameManager.animationManager.startAnimation(
      this.id + "_bounce",
      new SlideAnimation(
        this.gameManager,
        this.id + "_bounce",
        0.1,
        bounceAmplitudeX,
        bounceAmplitudeY,
        this.getAssets()
      )
    );
  }

  getOvershootAmount(target: number): number {
    if (!this.bounceEffect) return 0;
    return target * 0.1; // Overshoot by 10%
  }
}
